single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,749 @@
+        application in upgrade ?). Then we proceed to call the relation changed
+        handler and update the list of related hosts.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if self.state.upgrade_in_progress:
+            logger.warning(
+                "Adding replicas during an upgrade is not supported. The charm may be in a broken, unrecoverable state"
+            )
+            raise UpgradeInProgressError
+
+        self.on_relation_changed()
+        self.update_related_hosts()
+
+    def on_relation_changed(self) -> None:
+        """Handle relation changed events.
+
+        Adds the unit as a replica to the MongoDB replica set.
+        """
+        # Changing the monitor or the backup password will lead to non-leader
+        # units receiving a relation changed event. We must update the monitor
+        # and pbm URI if the password changes so that COS/pbm can continue to
+        # work
+        self.mongodb_exporter_config_manager.connect()
+        self.backup_manager.connect()
+
+        # only leader should configure replica set and we should do it only if
+        # the replica set is initialised.
+        if not self.charm.unit.is_leader() or not self.state.db_initialised:
+            return
+
+        try:
+            # Adds the newly added/updated units.
+            self.mongo_manager.process_added_units()
+        except (NotReadyError, PyMongoError) as e:
+            logger.error(f"Not reconfiguring: error={e}")
+            self.charm.status_manager.to_waiting("waiting to reconfigure replica set")
+            raise
+        self.charm.status_manager.to_active(None)
+
+    @override
+    def on_secret_changed(self, secret_label: str, secret_id: str) -> None:
+        """Handles secrets changes event.
+
+        When user run set-password action, juju leader changes the password inside the database
+        and inside the secret object. This action runs the restart for monitoring tool and
+        for backup tool on non-leader units to keep them working with MongoDB. The same workflow
+        occurs on TLS certs change.
+        """
+        if (
+            generate_secret_label(self.charm.app.name, self.charm.peer_rel_name, Scope.APP)
+            == secret_label
+        ):
+            scope = Scope.APP
+        elif (
+            generate_secret_label(self.charm.app.name, self.charm.peer_rel_name, Scope.UNIT)
+            == secret_label
+        ):
+            scope = Scope.UNIT
+        else:
+            logging.debug("Secret %s changed, but it's unknown", secret_id)
+            return
+        logging.debug("Secret %s for scope %s changed, refreshing", secret_id, scope)
+        self.state.secrets.get(scope)
+
+        # Always update the PBM and mongodb exporter configuration so that if
+        # the secret changed, the configuration is updated and will still work
+        # afterwards.
```

Comment:
question: what happens when a backup is in progress?

Reply:
That is a very good question and I don't have the answer and I'll definitely check that.
I think if a backup is in progress we should not try to change/update passwords.